linked_list/linked_list.py

```diff
@@ -86,32 +163,84 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)?
+    # Space Complexity: O(1)?
     def reverse(self):
-        pass
+        prev = None
+        current = self.head
+        while current is not None:
+            next = current.next
+            current.next = prev
+            prev = current
+            current = next
+        self.head = prev
   
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
+    # find middle value of a singly linked list
     def find_middle_value(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        count = 0
+        while current:
+            count += 1
+            current = current.next
+
+        current = self.head
+        for i in range(count // 2):
+            current = current.next
+
+        return current.value
+        
+
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     # Time Complexity: ?
     # Space Complexity: ?
+    # find nth node from the end of the linked list
     def find_nth_from_end(self, n):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+        count = 0
+        while current:
+            count += 1
+            current = current.next
+
+        current = self.head
+        for i in range(count - n):
+            current = current.next
+
+        return current.value
+
+        
+        
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     # Time Complexity: ?
     # Space Complexity: ?

```

Comment:
What do you think the space and time complexity of this is? Is the time complexity dependent on the length of the linked list? Is there any new variables being created that grow larger depending upon the input?